[CALCITE-6171] Support Gremlin Adapter

kaori-seasons opened the pull request:
For the specific discussion thread about this PR, please see: https://lists.apache.org/thread/s0cg1v0fqg84wxnq0gb0kyrb5074gxjh

Background See [CALCITE-6171](https://issues.apache.org/jira/browse/CALCITE-6171)

1. Parse the select/group by/having/order by in the node into the traversal of Gremlin syntax 
2. Traverse the method to complete the generation of the scanning method
3. After generating the scanning method based on the syntax node, generate the project traversal steps based on the specified label.
4. Rewrite the filter and scan in the execution plan and push them down
5. When completing a multi-table join, pre-determine whether the table fragment data hits the target and whether the expected data can be returned under the conditions of the scanning method.


I am currently improving the relevant integration tests for it. If you have any questions about this PR, friends in the community are welcome to give relevant suggestions.

kaori-seasons committed:
support gremlin query

kaori-seasons commented:
@asolimando @libenchao 

During this PR release, I noticed that you gave relevant suggestions on the mailing list. Can you take the time to review the PR?

JiajunBernoulli commented on settings.gradle.kts:
```diff
@@ -64,6 +64,7 @@ include(
     "cassandra",
     "core",
     "druid",
+    "gremlin",
```
alphabetic order

kaori-seasons replied:
solved

JiajunBernoulli commented:
@kaori-seasons 
Thank you for writing so much code.

Please fix CI and add document in site like [kafka_adapter](https://github.com/apache/calcite/blob/main/site/_docs/kafka_adapter.md).
> That will help others review your code.

kaori-seasons commented:
> @kaori-seasons Thank you for writing so much code.
> 
> Please fix CI and add document in site like [kafka_adapter](https://github.com/apache/calcite/blob/main/site/_docs/kafka_adapter.md).
> 
> > That will help others review your code.

Hello, sorry for the late reply, thank you for your reminder. There are some things that need to be dealt with. I'll fix the ci at the start of Friday.

XuQianJin-Stars commented:
Thank you for this feature. please check the ci pass.

kaori-seasons committed:
add sqlline definition

kaori-seasons commented:
> @kaori-seasons Thank you for writing so much code.
> 
> Please fix CI and add document in site like [kafka_adapter](https://github.com/apache/calcite/blob/main/site/_docs/kafka_adapter.md).
> 
> > That will help others review your code.

@JiajunBernoulli 
Hello, according to your instructions, I am supplementing the sqlline used by Gremlin Adapter.
But there is a problem in gremlin's native driver. This problem is that every time the user sends the query sql of the point table and edge table to the Gremlinserver query server to complete the query, but the returned context needs to include GraphTraversalSource or TinkerGraph.
Only when any object is used as a data source can we perform graph traversal operations based on this, as shown below

```
Object result = acceptor.connect(Arrays.asList(Storage.toPath(ConfigureSupport.generateTempFileFromResource(this.getClass(), "gremlin-server-integration.yaml", ".tmp")), "session-managed"));
      TinkerGraph graph = (TinkerGraph) result;
      GraphTraversalSource g = graph.traversal();
      g.addVertex('name','stephen');g.addVertex('name','marko');g.traversal().V()"

```

This means that I need to implement a new driver, which is the GremlinRemoteDriver in the code listing, to obtain the GraphTraversalSource or TinkerGraph object.
This work will take up more time, I will continue to complete this PR after it is ready.

kaori-seasons committed:
add sqlline test

kaori-seasons committed:
add sqlline test

kaori-seasons committed:
fix sqlline test

kaori-seasons committed:
adjust order

kaori-seasons committed:
fix ci

kaori-seasons commented:
@JiajunBernoulli 
Hello, I have fixed possible errors related to ci according to calcite's documentation, but when I run ci, I found the following error.

In fact, in the root file of gradle, I have added the gremlin module to module management. Since I am not familiar with gradle, can you please provide me with some guidance?


```
Could not determine the dependencies of task ':gremlin:forbiddenApisTest'.
> Could not resolve all task dependencies for configuration ':gremlin:testCompileClasspath'.
   > Could not resolve project :core.
     Required by:
         project :gremlin
      > Project :gremlin declares a dependency from configuration 
```
